### ISSUE TITLE
your email is Undeliverable

### ISSUE THREAD

**Issue by dhlee347** [nanazhu/Pseudo-Label-for-Deep-Neural-Networks#1]
You would get rid of 'g' in front of some functions in Matlab.. 

I'll re-implement the code soon.. 

function [trainData, trainLabels, unsupData, unsupLabels, valData, valLabels] = splitData(data, labels, nTrain, nVal, nUnsup)

[trainData, trainLabels, data, labels] = pickEqually(data, labels, nTrain);
[valData,   valLabels,   data, labels] = pickEqually(data, labels, nVal);
if nUnsup > length(labels) || nUnsup == 0, nUnsup = length(labels); end
unsupData = data(:,1:nUnsup);
unsupLabels = labels(1:nUnsup);

end

function [pData, pLabels, rData, rLabels] = pickEqually(data, labels, N)
% N must be dividable by nLabels

nLabels = max(labels);
nData = length(labels);
rp = randperm(nData);

data = data(:,rp); rData = data;
labels = labels(rp); rLabels = labels;

pIdx = [];
for l = 1:nLabels
    idx = find(labels==l);
    pIdx = [pIdx; idx(1:N/nLabels)];
end

pData = data(:,pIdx);
pLabels = labels(pIdx);

rData(:,pIdx) = [];
rLabels(pIdx) = [];

end

function [pData, pLabels, rData, rLabels] = pickRandomly(data, labels, N)

nData = length(labels);
rp = randperm(nData);

data = data(:,rp); 
labels = labels(rp);

pData = data(:,1:N);
pLabels = labels(1:N);

rData = data(:,N+1:end);
rLabels = labels(N+1:end);

end

clear all; fid = fopen('log_1015.txt','w'); % log file

%%%% utility functions %%%%

e = 1e-8; % infinitesimal number for preventing divide-by-zero error
infer = @(W,B,V) bsxfun(@plus, W*V, B); % feed-forward inference of neural network layer
relu  = @(x) max(0,x);  d_relu = @(x)  x > 0; % rectified linear activation function & the derivative of that
sigm  = @(x) 1./(1+exp(-x));  d_sigm = @(x) x.*(1-x); % sigmoid activation function and the derivative of that
CE    = @(x,y) -( x.*log(y+e) + (1-x).*log(1-y+e) ); % sigmoid activation function and the derivative of that
d_CE  = @(x,y) - x./(y+e) + (1-x)./(1-y+e); % sigmoid activation function and the derivative of that
drop  = @(s,p) s.*(grand(size(s)) > p); % sigmoid activation function and the derivative of that

alpha = @(x) 3*min(max(x-100,0)/(600-100),1); % x-100 -> x-200 for 100 labeled data % for weight regularization in dropout - not used

% data   - the number of dimension (784)  X  the number of samples (60000) (column major) (value 1 ~ 255)
% labels - the number of samples  X  1 (value - 1 ~ 10)
% MNIST.mat must have trainData, trainLabels, testData, testLabels
load MNIST; data = trainData/255; labels = trainLabels;

for nTrain = 600 % the number of training samples
    for r=1:10 % the number of trials 
        
        %%%% prepare split data in a semi-supervised setting %%%%
        
        [trainData, trainLabels, unsupData, unsupLabels, valData, valLabels] = splitData(data, labels, nTrain, 1000, 60000-nTrain-1000);
        trainTargets = gsingle(full( sparse(1:size(trainLabels,1), double(trainLabels), 1) )'); % get 1-of-K code of labels

        %%%% the size of networks %%%%
        
        nV=784; nH1=5000; nC=10; nH2=nC; 
        
        %%%% initialize %%%%
        
        W1 = 0.01*grandn(nH1, nV);  B1 = gzeros(size(W1,1),1); % initialize the weigths randomly
        W2 = 0.01*grandn(nH2,nH1);  B2 = gzeros(size(W2,1),1); 
        
        dW1=W1; dW1(:)=0; dB1=B1; dB1(:)=0; dW2=W2; dW2(:)=0; dB2=B2; dB2(:)=0; % initialize delta W and B with zeros

        %%%% hyper-parameters %%%%
        
        M = size(trainData,2); uM = size(unsupData,2); m=32; um=256; % m - mini-batch size of labeled data, um - mini-batch size of unlabeled data
        maxIter=3000; eta=1.5; moi=0.5; mof=0.99; mo = moi; rng('shuffle'); tic; % eta - initial learning rate, mo - momentum, moi - initial momentum, mof - final momentum
        
        %%%% main iteration loop of Stocastic Gradient Descent with Dropout %%%%
        
        for iter = 1:maxIter
            scost = zeros(floor(M/m),1); ucost = scost;
            
            %%%% sub loop for mini-batches %%%%
            
            for i = 1:floor(M/m)
                mi = randi(M,m,1); umi = randi(uM,um,1); % mini-batch indexes of labeled data and unlabeled data

                %%%% inference %%%%
                
                % labeled data and unlabeled data
                V  = drop(gsingle([trainData(:,mi) unsupData(:,umi)]),0.2); % mini-batch indexes of labeled data and unlabeled data
                H1 = drop(relu(infer(W1,B1,V)), 0.5); % calculate 1st hidden units with rectified linear activation function
                H2 = sigm(infer(W2,B2,H1)); % calculate label probabilities
                
                sP = H2(:,1:m); % sP - predicted probabilities of labeled data
                uP = H2(:,m+1:end); % uP - predicted probabilities of unlabeled data

                sT = trainTargets(:,mi); % sT - the targets of labeled data
                uP_= sigm(infer(W2,B2, 0.5*relu(infer(W1,B1,gsingle(unsupData(:,umi)))))); % no dropout for pseudo-label
                uT = bsxfun(@eq, uP_, max(uP_)); % uT - the targets of unlabeled data : pseudo-Label
                
                %%%% costs & gradient %%%%
                
                scost(i) = mean(sum(CE(sT,sP))); % calculate costs for labeled data and unlabeled data
                ucost(i) = mean(sum(CE(uT,uP)));
               
                dH2 = [d_CE(sT,sP)/m  alpha(iter)*d_CE(uT,uP)/um].*d_sigm(H2); % the derivate of 2nd hidden units
                dH1 = W2'*dH2.*d_relu(H1); % the derivate of 1st hidden units
                
                gW2 = dH2*H1'; gB2 = sum(dH2,2); gW1 = dH1*V'; gB1 = sum(dH1,2); % gradients of weights and biases
                
                dW2 = mo*dW2 - (1-mo)*eta*gW2;  W2=W2+dW2;    dB2 = mo*dB2 - (1-mo)*eta*gB2;  B2=B2+dB2; % increments of weights and biases using momentum
                dW1 = mo*dW1 - (1-mo)*eta*gW1;  W1=W1+dW1;    dB1 = mo*dB1 - (1-mo)*eta*gB1;  B1=B1+dB1;
            end
            
            eta = eta*0.998; % exponentially decaying learning rate (from dropout paper)
            mo = min(1,iter/500)*mof + max(0,1-iter/500)*moi; % linearly increasing momentum : moi ~ mof (from dropout paper)
            
            %%%% test networks %%%%
            
            if mod(iter,100)==0,
                predict = @(x) max(infer(W2,B2, 0.5*relu(infer(W1,B1,x)) )); % predict labels with the whole network
                getErr = @(x,y) 100*( 1 - sum( x' == gsingle(y) ) / length(x) ); % get prediction error
                
                [~,predValLabels ] = predict(valData);  valErr  = getErr(predValLabels, valLabels);
                [~,predTestLabels] = predict(testData); testErr = getErr(predTestLabels,testLabels);
                
                msg = sprintf('>%d[%4d/%4d] cost=%f/%f Err=%.2f/%.2f - %f sec\n',nTrain,iter,maxIter,mean(scost),mean(ucost),valErr,testErr,toc); 
                fprintf(msg); fprintf(fid,msg);
                %displayNetwork(W1(1:144,:)');
            end
        end
    fprintf('\n');        
    end
    fprintf('--------------\n\n');
end
